README.md
# Indicator Button

A Flutter Button with indicates progress

![Pub](https://img.shields.io/pub/v/indicator_button) ![GitHub](https://img.shields.io/github/license/huobazi/indicator_button)

## Demo:

![Demo 1](https://github.com/huobazi/indicator_button/blob/master/demo1.gif?raw=true)

![Demo 2](https://github.com/huobazi/indicator_button/blob/master/demo2.gif?raw=true)

![Demo 3](https://github.com/huobazi/indicator_button/blob/master/demo3.gif?raw=true)

## Getting Started

- Add dependency to your pubspec.yaml
```yaml
dependencies:
  indicator_button: ^1.0.1
```

- Install packages

```bash
flutter pub get
```

- Import it
```dart
import 'package:indicator_button/indicator_button.dart';
```

- Add indicator to your widget tree
```dart
  int _targetUserId = 123;
  List<int> _followingUserIdList = [123];

  Widget _createFollowButton() {
    var followed = _followingUserIdList.contains(_targetUserId);
    var buttonText = followed ? 'Following' : ' +  Follow';
    var buttonBackgroundColor = followed ? Colors.white30 : Colors.white;
    var borderColor = followed ? Colors.grey : Colors.blue;
    var textColor = followed ? Colors.grey : Colors.blue;

    var button = IndicatorButton(
      backgroundColor: buttonBackgroundColor,
      progressColor: textColor,
      borderColor: borderColor,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: <Widget>[
          Icon(
            Icons.remove_red_eye,
            size: 20.0,
            color: textColor,
          ),
          Text(
            buttonText,
            style: TextStyle(
              fontWeight: FontWeight.bold,
              fontSize: 15.0,
              color: textColor,
            ),
          ),
        ],
      ),
      onPressed: () async {
        await Future.delayed(new Duration(milliseconds: 2000), () {
          setState(() {
            if (followed) {
              _followingUserIdList.remove(_targetUserId);
            } else {
              _followingUserIdList.add(_targetUserId);
            }
          });
        });
      },
    );

    return Container(
      width: 130.0,
      height: 32.0,
      child: button,
    );
  }
```

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Contributors

See the [Contributors List](https://github.com/huobazi/indicator_button/graphs/contributors).

## CHANGE LOG

See the [CHANGELOGS.md](https://github.com/huobazi/indicator_button/blob/master/CHANGELOG.md).


## License

This project is licensed under the MIT License - see the [LICENSE](https://github.com/huobazi/indicator_button/blob/master/LICENSE) file for details




CHANGELOG.md
## [1.0.1] - 2019-08-02
* add example doc
## [1.0.0] - 2019-08-02
* add example and english demo
## [0.0.3] - 2019-08-02
* fix demo.
## [0.0.2] - 2019-08-02
* fix bug and readme.
## [0.0.1] - 2019-07-31.
* first release.


test/indicator_button_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:indicator_button/indicator_button.dart';

void main() {
}

lib/indicator_button.dart
library indicator_button;


import 'package:flutter/material.dart';

typedef FutureVoidCallback = Future<void> Function();

/// A Flutter Button with indicates progress
class IndicatorButton extends StatefulWidget {
  final FutureVoidCallback onPressed;
  final Widget child;
  final Color backgroundColor;
  final Color progressColor;
  final Color borderColor;

  IndicatorButton({
    Key key,
    @required this.onPressed,
    this.child,
    this.backgroundColor,
    this.progressColor,
    this.borderColor,
  })
      : assert(onPressed != null),
        super(key: key);

  @override
  _IndicatorButtonState createState() => _IndicatorButtonState();
}

class _IndicatorButtonState extends State<IndicatorButton>
    with TickerProviderStateMixin {
  AnimationController _animationController;
  Animation<Offset> _animation;
  bool _progressing = false;

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 350),
    );
    _animation = TweenSequence<Offset>([
      TweenSequenceItem<Offset>(
          tween: Tween(begin: Offset(0, 0), end: Offset(0.03, 0)), weight: 1),
      TweenSequenceItem<Offset>(
          tween: Tween(begin: Offset(0.03, 0), end: Offset(-0.03, 0)),
          weight: 2),
      TweenSequenceItem<Offset>(
          tween: Tween(begin: Offset(-0.03, 0), end: Offset(0.03, 0)),
          weight: 2),
      TweenSequenceItem<Offset>(
          tween: Tween(begin: Offset(0.03, 0), end: Offset(-0.03, 0)),
          weight: 2),
      TweenSequenceItem<Offset>(
          tween: Tween(begin: Offset(-0.03, 0), end: Offset(0, 0)), weight: 1)
    ]).animate(CurvedAnimation(
      parent: _animationController,
      curve: Curves.linear,
    ));
  }

  @override
  Widget build(BuildContext context) {
    Color backgroundColor =
        widget.backgroundColor ?? Theme
            .of(context)
            .primaryColor;
    Color borderColor = widget.borderColor ?? Theme
        .of(context)
        .primaryColor;
    Color progressColor = widget.progressColor ?? Colors.white;

    return AnimatedBuilder(
      animation: _animationController,
      builder: (context, child) {
        return SlideTransition(
          position: _animation,
          child: LayoutBuilder(
            builder: (context, constraints) {
              var buttonHeight = constraints.maxHeight;
              if (buttonHeight == double.infinity) {
                buttonHeight = 40;
              }
              Widget buttonContent = widget.child ?? Container();
              if (_progressing) {
                buttonContent = SizedBox(
                  height: buttonHeight,
                  width: buttonHeight,
                  child: Padding(
                    padding: EdgeInsets.all(8),
                    child: CircularProgressIndicator(
                      valueColor: AlwaysStoppedAnimation<Color>(
                          progressColor ?? Colors.white),
                      strokeWidth: 3,
                    ),
                  ),
                );
              }

              return AnimatedBuilder(
                animation: _animationController,
                builder: (context, child) {
                  return FlatButton(
                    child: buttonContent,
                    onPressed: () async {
                      setState(
                            () {
                          _progressing = true;
                        },
                      );
                      await widget.onPressed();
                      setState(
                            () {
                          _progressing = false;
                        },
                      );
                    },
                    color: backgroundColor,
                    shape: Border.all(color: borderColor),
                  );
                },
              );
            },
          ),
        );
      },
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:indicator_button/indicator_button.dart';
import 'dart:async';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Indicator Button Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Indicator Button Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'Tap the button:',
            ),
            SizedBox(
              height: 30,
            ),
            _createFollowButton(),
          ],
        ),
      ),
    );
  }

  int _targetUserId = 123;
  List<int> _followingUserIdList = [123];

  Widget _createFollowButton() {
    var followed = _followingUserIdList.contains(_targetUserId);
    var buttonText = followed ? 'Following' : ' +  Follow';
    var buttonBackgroundColor = followed ? Colors.white30 : Colors.white;
    var borderColor = followed ? Colors.grey : Colors.blue;
    var textColor = followed ? Colors.grey : Colors.blue;

    var button = IndicatorButton(
      backgroundColor: buttonBackgroundColor,
      progressColor: textColor,
      borderColor: borderColor,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: <Widget>[
          Icon(
            Icons.remove_red_eye,
            size: 20.0,
            color: textColor,
          ),
          Text(
            buttonText,
            style: TextStyle(
              fontWeight: FontWeight.bold,
              fontSize: 15.0,
              color: textColor,
            ),
          ),
        ],
      ),
      onPressed: () async {
        await Future.delayed(new Duration(milliseconds: 2000), () {
          setState(() {
            if (followed) {
              _followingUserIdList.remove(_targetUserId);
            } else {
              _followingUserIdList.add(_targetUserId);
            }
          });
        });
      },
    );

    return Container(
      width: 130.0,
      height: 32.0,
      child: button,
    );
  }
}
